### ISSUE TITLE
Support OpenShift usage for Camel K extension

### ISSUE THREAD

**Issue by apupier** [eclipse-che/che#16131]
### Is your enhancement related to a problem? Please describe.

Currently, the Camel K plugin is coming with Kubernetes plugin. Camel K can be deployed to OpenShift too. it would nice to provide OpenShift connector plugin too.

### Describe the solution you'd like

Several possibilities:
- Che can better support VS Code extensions that have dependencies between them in separate definitions. So that we can include Kubernetes/OpenShift Connector and Camel K separately and they can communicate between them and share the credentials and some critical configuration folder
- include OpenShift Connector in the dependencies of Camel K plugin. it means to enhance the sidecar too. Would be nice if the OpenShift sidecar is inheriting from Kubernetes one so that siecar Camel can inherit from OpenShift one instead of Kubernetes directly and don't duplicate code from one of them. see also https://github.com/eclipse/che/issues/16130

**Comment by apupier** [comment]
provided a quick and dirty proof of concept:
https://raw.githubusercontent.com/apupier/che-devfile-registry/camlekwithopenshift/devfiles/apache-camel-k/devfile.yaml
https://raw.githubusercontent.com/apupier/che-devfile-registry/camlekwithopenshift/devfiles/apache-camel-k/devfile.yaml
https://raw.githubusercontent.com/apupier/che-plugin-registry/camelkwithopenshift/v3/plugins/redhat/vscode-camelk/0.0.12/meta.yaml
https://github.com/apupier/che-sidecar-camelk/blob/0.0.12-withOC/Dockerfile

**Comment by l0rd** [comment]
@apupier I would consider this issue as a duplicate of https://github.com/eclipse/che/issues/15966

**Comment by tsmaeder** [comment]
Isnt' there a separate OpenShift Connector plugin in the plugin registry already?

**Comment by apupier** [comment]
> I would consider this issue a duplicate of #15966 (and @benoitf is currently working on it). Am I correct?

I think the #15966 will solve a part of issue in a more generic way and that will be awesome.

Based on the description, I'm not sure it is not covering a shared login mechanism to Kubernetes/OpenShift and the shared access to CLI tools.

**Comment by apupier** [comment]
> Isnt' there a separate OpenShift Connector plugin in the plugin registry already? I mean why do we need to include the OpenShift Connector extension in the Camel K plugin and not just use the existing one in the devfile?

- the login done in OpenShift Connector container is not reachable by Camel K
- it is convenient to have access to all CLIs tools on the same container (oc/kamel)

**Comment by tsmaeder** [comment]
> 
>     * the login done in OpenShift Connector container is not reachable by Camel K
I see that, but I guess sharing env between plugins is another problem we need to solve in our workspace model.

> 
>     * it is convenient to have access to all CLIs tools on the same container (oc/kamel)
The plugin sidecar should serve the Theia plugins, not the CLI. That's what the dev container is for.

**Comment by apupier** [comment]
> > ```
> > * it is convenient to have access to all CLIs tools on the same container (oc/kamel)
> > ```
> 
> 
> The plugin sidecar should serve the Theia plugins, not the CLI. That's what the dev container is for.

interesting, I miss that.
In this specific case, the VS Code extension requires the binaries on the CLI. So providing it in the sidecar container avoids to duplicate the binaries in both container. Or maybe the sidecar can access the "dev container"?
How can we configure the "dev container" with devfile? I found no reference in the [doc](https://redhat-developer.github.io/devfile/devfile).

**Comment by tsmaeder** [comment]
> So providing it in the sidecar container avoids to duplicate the binaries in both container. 
No, it doesn't in general: the extension needs to bring it's own prerequisites, this might mean a different version of a tool than the one the user targets.

> Or maybe the sidecar can access the "dev container"?

No, that's not the workspace model as I understand it.

**Comment by ericwill** [comment]
I think this is partially solved, as we have the [che-container-tools](https://github.com/che-dockerfiles/che-container-tools) base image now. If you need container related tooling you can probably use that sidecar as a base image, and then you'll get lots of the deps you need for free.
